sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Main.scala

```diff
@@ -65,20 +93,61 @@ object Main extends IOApp {
+        if (List(config.withRaw, config.withEnrichedTsv, config.withEnrichedJson).count(identity) > 1)
+          throw new RuntimeException(s"Kinesis could only output in single format")
+        if (config.compress) {
+          throw new RuntimeException(s"Kinesis doesn't support compression")
+        }
+        val kinesisClient: KinesisAsyncClient = KinesisClientUtil.createKinesisAsyncClient(KinesisAsyncClient.builder())
+
+        def reqBuilder(event: Event): PutRecordRequest = PutRecordRequest.builder()
+          .streamName(outputDir.getRawAuthority)
+          .partitionKey(event.event_id.toString)
+          .data(SdkBytes.fromUtf8String(event.toTsv))
+          .build()
+
+        st: Stream[F, GenOutput] =>
+          st.map(_._2)
+            .flatMap(Stream.emits)
+            .map(reqBuilder)
+            .parEvalMap(10)(e =>
+              Async[F].fromCompletableFuture(Sync[F].delay(kinesisClient.putRecord(e)))
+            ).void
+      } else {
+        RotatingSink.rotate(config.payloadsPerFile) { idx =>
+          val pipe1: Pipe[F, GenOutput, INothing] = _.map(_._1)
+            .through(Serializers.rawSerializer(config.compress))
+            .through(sinkFor("raw", idx, config.withRaw))
+          val pipe2: Pipe[F, GenOutput, INothing] = _.flatMap(in => Stream.emits(in._2))
+            .through(Serializers.enrichedTsvSerializer(config.compress))
+            .through(sinkFor("enriched", idx, config.withEnrichedTsv))
+          val pipe3: Pipe[F, GenOutput, INothing] = _.flatMap(in => Stream.emits(in._2))
+            .through(Serializers.enrichedJsonSerializer(config.compress))
+            .through(sinkFor("transformed", idx, config.withEnrichedJson))
+          in: Stream[F, GenOutput] => in.broadcastThrough(pipe1, pipe2, pipe3)
+        }
+      }
+
     eventStream
       .take(config.payloadsTotal.toLong)
-      .through(RotatingSink.rotate(config.payloadsPerFile) { idx =>
-        val pipe1: Pipe[F, GenOutput, INothing]  = _.map(_._1)
-          .through(Serializers.rawSerializer(config.compress))
-          .through(sinkFor("raw", idx, config.withRaw))
-        val pipe2: Pipe[F, GenOutput, INothing]  = _.flatMap(in => Stream.emits(in._2))
-          .through(Serializers.enrichedTsvSerializer(config.compress))
-          .through(sinkFor("enriched", idx, config.withEnrichedTsv))
-        val pipe3: Pipe[F, GenOutput, INothing]  = _.flatMap(in => Stream.emits(in._2))
-          .through(Serializers.enrichedJsonSerializer(config.compress))
-          .through(sinkFor("transformed", idx, config.withEnrichedJson))
-        in: Stream[F, GenOutput] => in.broadcastThrough(pipe1, pipe2, pipe3)
-      })
+      .zipWithScan1((0, 0)) { case (idx, el) => (el._2.length + idx._1, 1 + idx._2) }
+      .evalTap {
+        case (_, idx) =>
+          if (idx._1 % 10000 == 0 && idx._1 != 0) {
+            Sync[F].delay(println(s"processed events: ${idx._1}..."))
+          } else if (idx._2 == config.payloadsTotal.toLong) {
```

Comment:
This `else` doesn't look right to me.  I would guess if `payloadsTotal` is exactly equal to 10000 then you want to print the second block, not the first block.

But maybe I haven't understood what you are trying to print and when.